executeCommand changeSet fails on >=3.10.1

<!--- This environment context section helps us quickly review your PR. 
      Please take a minute to fill-out this information. -->

## Environment
**Liquibase Version**: >=3.10.1

**Liquibase Integration & Version**: ant

**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**: MariaDB 10.3

**Operating System Type & Version**: Debian Stretch

## Description
A changeSet with only a `<executeCommand>` statement fails in >=3.10.1, it works under 3.10.0.

```
Unable to update database. liquibase.exception.MigrationFailedException: Migration failed for change set liquibase.xml::xxxx::xxxx:
     Reason: java.lang.ClassCastException: liquibase.database.jvm.JdbcConnection cannot be cast to com.datical.liquibase.ext.database.jvm.ProJdbcConnection
        at liquibase.changelog.ChangeSet.execute(ChangeSet.java:659)
        at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:53)
        at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:97)
        at liquibase.Liquibase.update(Liquibase.java:201)
        at liquibase.Liquibase.update(Liquibase.java:178)
        at liquibase.Liquibase.update(Liquibase.java:328)
        at liquibase.integration.ant.DatabaseUpdateTask.executeWithLiquibaseClassloader(DatabaseUpdateTask.java:34)
        at liquibase.integration.ant.BaseLiquibaseTask.execute(BaseLiquibaseTask.java:88)
        at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:293)
        at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
        at org.apache.tools.ant.Task.perform(Task.java:348)
        at org.apache.tools.ant.Target.execute(Target.java:435)
        at org.apache.tools.ant.Target.performTasks(Target.java:456)
        at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1405)
        at org.apache.tools.ant.helper.SingleCheckExecutor.executeTargets(SingleCheckExecutor.java:38)
        at org.apache.tools.ant.Project.executeTargets(Project.java:1260)
        at org.apache.tools.ant.taskdefs.Ant.execute(Ant.java:441)
        at org.apache.tools.ant.taskdefs.CallTarget.execute(CallTarget.java:105)
        at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:293)
        at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
        at org.apache.tools.ant.Task.perform(Task.java:348)
        at org.apache.tools.ant.Target.execute(Target.java:435)
        at org.apache.tools.ant.Target.performTasks(Target.java:456)
        at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1405)
        at org.apache.tools.ant.Project.executeTarget(Project.java:1376)
        at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)
        at org.apache.tools.ant.Project.executeTargets(Project.java:1260)
        at org.apache.tools.ant.Main.runBuild(Main.java:857)
        at org.apache.tools.ant.Main.startAnt(Main.java:236)
        at org.apache.tools.ant.launch.Launcher.run(Launcher.java:287)
        at org.apache.tools.ant.launch.Launcher.main(Launcher.java:113)
Caused by: java.lang.ClassCastException: liquibase.database.jvm.JdbcConnection cannot be cast to com.datical.liquibase.ext.database.jvm.ProJdbcConnection
        at com.datical.liquibase.ext.SqlPlusRunner.buildConnectionString(Unknown Source)
        at com.datical.liquibase.ext.SqlPlusRunner.createFinalCommandArray(Unknown Source)
        at liquibase.change.core.ExecuteShellCommandChange.generateStatements(ExecuteShellCommandChange.java:142)
        at liquibase.database.AbstractJdbcDatabase.executeStatements(AbstractJdbcDatabase.java:1256)
        at liquibase.changelog.ChangeSet.execute(ChangeSet.java:622)
        ... 36 more
```



┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-637) by [Unito](https://www.unito.io/learn-more)

Thanks for this info! It's helped me learn a bit more. I've downloaded the same jar you did and examined its contents. I found the pro classes contained in the jar. I had no idea they were there so I definitely learned something. 

Here's what I've figured out about the error:

```
Caused by: java.lang.ClassCastException: liquibase.database.jvm.JdbcConnection cannot be cast to com.datical.liquibase.ext.database.jvm.ProJdbcConnection
        at com.datical.liquibase.ext.SqlPlusRunner.buildConnectionString(Unknown Source)
        at com.datical.liquibase.ext.SqlPlusRunner.createFinalCommandArray(Unknown Source)
        at liquibase.change.core.ExecuteShellCommandChange.generateStatements(ExecuteShellCommandChange.java:142)
```

When I look at the `ExecuteShellComamndChange` class I can clearly see it's designed to be subclassed and that is exactly what `SqlPlusRunner` does. Now I don't entirely know why `SqlPlusRunner` is selected over the `ExecuteShellComamndChange` class but I think it probably has something to do with Liquibase's priority system for service loading. The `ExecuteShellComamndChange` class has a default priority but I don't know what `SqlPlusRunner`'s priority is as I don't have the source code for pro. Since the `SqlPlusRunner` is clearly selected we can assume it's priority is higher.

It's the `SqlPlusRunner` class that tries to cast the `JdbcConnection` to a `ProJdbcConnection` which is definitely not possible right now in the ant tasks as they are not aware of pro classes and don't create them.

One possible way to fix this would be to check to see if the class can be cast before performing the cast. Another possible fix is to not load the pro classes when pro is not requested. Another option is to enhance the Ant tasks to be pro-aware (have some sort of boolean flag that can be toggled). All of these fixes are outside of my abilities as I don't have access to the pro codebase.

And all of this is even more confusing since you say that it sometimes works! That one I'm still baffled by. My only hunch on that is the nested `<antcalls>` are messing with the classpath order and classloaders so that _sometimes_ the pro classes are loaded later and therefore not found so Liquibase can select the regular `ExecuteShellComamndChange` class.

@molivasdat Do you have any thoughts on this?

--- a/liquibase-core/src/main/java/liquibase/database/DatabaseFactory.java
+++ b/liquibase-core/src/main/java/liquibase/database/DatabaseFactory.java
@@ -2,7 +2,6 @@ package liquibase.database;
 
 import liquibase.Scope;
 import liquibase.database.core.UnsupportedDatabase;
-import liquibase.database.jvm.JdbcConnection;
 import liquibase.exception.DatabaseException;
 import liquibase.exception.UnexpectedLiquibaseException;
 import liquibase.logging.Logger;
@@ -11,7 +10,6 @@ import liquibase.util.StringUtil;
 
 import java.io.File;
 import java.io.FileInputStream;
-import java.sql.Connection;
 import java.sql.Driver;
 import java.util.*;
 
@@ -165,10 +163,6 @@ public class DatabaseFactory {
         }
 
         DatabaseConnection databaseConnection = null;
-        driver = StringUtil.trimToNull(driver);
-        if (driver == null) {
-            driver = DatabaseFactory.getInstance().findDefaultDriver(url);
-        }
 
         try {
             Driver driverObject;
@@ -178,16 +172,17 @@ public class DatabaseFactory {
                 databaseFactory.register((Database) Class.forName(databaseClass, true, Scope.getCurrentScope().getClassLoader()).getConstructor().newInstance());
             }
 
+            String selectedDriverClass = StringUtil.trimToNull(driver);
+            if (selectedDriverClass == null) {
+                selectedDriverClass = databaseFactory.findDefaultDriver(url);
+            }
+
+            if (selectedDriverClass == null) {
+                throw new RuntimeException("Driver class was not specified and could not be determined from the url (" + url + ")");
+            }
+            
             try {
-                if (driver == null) {
-                    driver = databaseFactory.findDefaultDriver(url);
-                }
-
-                if (driver == null) {
-                    throw new RuntimeException("Driver class was not specified and could not be determined from the url (" + url + ")");
-                }
-
-                driverObject = (Driver) Class.forName(driver, true, Scope.getCurrentScope().getClassLoader()).getConstructor().newInstance();
+                driverObject = (Driver) Class.forName(selectedDriverClass, true, Scope.getCurrentScope().getClassLoader()).getConstructor().newInstance();
             } catch (Exception e) {
                 throw new RuntimeException("Cannot find database driver: " + e.getMessage());
             }
@@ -237,9 +232,9 @@ public class DatabaseFactory {
                 }
             }
 
-            if(driver.contains("oracle")) {
+            if(selectedDriverClass.contains("oracle")) {
               driverProperties.put("remarksReporting", "true");
-            } else if(driver.contains("mysql")) {
+            } else if(selectedDriverClass.contains("mysql")) {
               driverProperties.put("useInformationSchema", "true");
             }
 

--- a/liquibase-core/src/main/java/liquibase/database/DatabaseFactory.java
+++ b/liquibase-core/src/main/java/liquibase/database/DatabaseFactory.java
@@ -140,6 +140,15 @@ public class DatabaseFactory {
         return this.findCorrectDatabaseImplementation(openConnection(url, username, password, driver, databaseClass, driverPropertiesFile, propertyProviderClass, resourceAccessor));
     }
 
+    public Database openDatabase(String url,
+                                 String username,
+                                 String driver,
+                                 String databaseClass,
+                                 Properties driverProperties,
+                                 ResourceAccessor resourceAccessor) throws DatabaseException {
+        return this.findCorrectDatabaseImplementation(openConnection(url, username, driver, databaseClass, driverProperties, resourceAccessor));
+    }
+
     public DatabaseConnection openConnection(String url,
                                              String username,
                                              String password,
@@ -254,10 +263,10 @@ public class DatabaseFactory {
         return selectedDriverClass;
     }
 
-    private Driver loadDriver(String selectedDriverClass) {
+    private Driver loadDriver(String driverClass) {
         Driver driverObject;
         try {
-            driverObject = (Driver) Class.forName(selectedDriverClass, true, Scope.getCurrentScope().getClassLoader()).getConstructor().newInstance();
+            driverObject = (Driver) Class.forName(driverClass, true, Scope.getCurrentScope().getClassLoader()).getConstructor().newInstance();
         } catch (Exception e) {
             throw new RuntimeException("Cannot find database driver: " + e.getMessage());
         }

--- a/liquibase-core/src/main/java/liquibase/database/DatabaseFactory.java
+++ b/liquibase-core/src/main/java/liquibase/database/DatabaseFactory.java
@@ -10,6 +10,7 @@ import liquibase.util.StringUtil;
 
 import java.io.File;
 import java.io.FileInputStream;
+import java.io.IOException;
 import java.sql.Driver;
 import java.util.*;
 
@@ -162,75 +163,22 @@ public class DatabaseFactory {
             return offlineConnection;
         }
 
-        DatabaseConnection databaseConnection = null;
-
+        DatabaseConnection databaseConnection;
         try {
-            Driver driverObject;
             DatabaseFactory databaseFactory = DatabaseFactory.getInstance();
             if (databaseClass != null) {
                 databaseFactory.clearRegistry();
                 databaseFactory.register((Database) Class.forName(databaseClass, true, Scope.getCurrentScope().getClassLoader()).getConstructor().newInstance());
             }
 
-            String selectedDriverClass = StringUtil.trimToNull(driver);
-            if (selectedDriverClass == null) {
-                selectedDriverClass = databaseFactory.findDefaultDriver(url);
-            }
-
-            if (selectedDriverClass == null) {
-                throw new RuntimeException("Driver class was not specified and could not be determined from the url (" + url + ")");
-            }
-            
-            try {
-                driverObject = (Driver) Class.forName(selectedDriverClass, true, Scope.getCurrentScope().getClassLoader()).getConstructor().newInstance();
-            } catch (Exception e) {
-                throw new RuntimeException("Cannot find database driver: " + e.getMessage());
-            }
+            String selectedDriverClass = findDriverClass(url, driver, databaseFactory);
+            Driver driverObject = loadDriver(selectedDriverClass);
 
             if (driverObject instanceof LiquibaseExtDriver) {
                 ((LiquibaseExtDriver)driverObject).setResourceAccessor(resourceAccessor);
             }
 
-            Properties driverProperties;
-            if (propertyProviderClass == null) {
-                driverProperties = new Properties();
-            } else {
-                driverProperties = (Properties) Class.forName(propertyProviderClass, true, Scope.getCurrentScope().getClassLoader()).getConstructor().newInstance();
-            }
-
-            if (username != null) {
-                driverProperties.put("user", username);
-            }
-            if (password != null) {
-                driverProperties.put("password", password);
-            }
-            if (null != driverPropertiesFile) {
-                File propertiesFile = new File(driverPropertiesFile);
-                if (propertiesFile.exists()) {
-                    LOG.fine(
-                            "Loading properties from the file:'" + driverPropertiesFile + "'"
-                    );
-                    FileInputStream inputStream = new FileInputStream(propertiesFile);
-                    try {
-                        driverProperties.load(inputStream);
-                    } finally {
-                        inputStream.close();
-                    }
-                } else {
-                    throw new RuntimeException("Can't open JDBC Driver specific properties from the file: '"
-                            + driverPropertiesFile + "'");
-                }
-            }
-
-
-            LOG.fine("Properties:");
-            for (Map.Entry entry : driverProperties.entrySet()) {
-                if (entry.getKey().toString().toLowerCase().contains("password")) {
-                    Scope.getCurrentScope().getLog(getClass()).fine("Key:'" + entry.getKey().toString() + "' Value:'**********'");
-                } else {
-                    LOG.fine("Key:'" + entry.getKey().toString() + "' Value:'" + entry.getValue().toString() + "'");
-                }
-            }
+            Properties driverProperties = buildDriverProperties(username, password, driverPropertiesFile, propertyProviderClass);
 
             if(selectedDriverClass.contains("oracle")) {
               driverProperties.put("remarksReporting", "true");
@@ -278,6 +226,72 @@ public class DatabaseFactory {
         }
         return implementedDatabases.get(shortName).iterator().next();
 
+    }
+
+    private String findDriverClass(String url, String driver, DatabaseFactory databaseFactory) {
+        String selectedDriverClass = StringUtil.trimToNull(driver);
+        if (selectedDriverClass == null) {
+            selectedDriverClass = databaseFactory.findDefaultDriver(url);
+        }
+
+        if (selectedDriverClass == null) {
+            throw new RuntimeException("Driver class was not specified and could not be determined from the url (" + url + ")");
+        }
+        return selectedDriverClass;
+    }
+
+    private Driver loadDriver(String selectedDriverClass) {
+        Driver driverObject;
+        try {
+            driverObject = (Driver) Class.forName(selectedDriverClass, true, Scope.getCurrentScope().getClassLoader()).getConstructor().newInstance();
+        } catch (Exception e) {
+            throw new RuntimeException("Cannot find database driver: " + e.getMessage());
+        }
+        return driverObject;
+    }
+
+    private Properties buildDriverProperties(String username, String password, String driverPropertiesFile, String propertyProviderClass) throws InstantiationException, IllegalAccessException, java.lang.reflect.InvocationTargetException, NoSuchMethodException, ClassNotFoundException, IOException {
+        Properties driverProperties;
+        if (propertyProviderClass == null) {
+            driverProperties = new Properties();
+        } else {
+            driverProperties = (Properties) Class.forName(propertyProviderClass, true, Scope.getCurrentScope().getClassLoader()).getConstructor().newInstance();
+        }
+
+        if (username != null) {
+            driverProperties.put("user", username);
+        }
+        if (password != null) {
+            driverProperties.put("password", password);
+        }
+        if (null != driverPropertiesFile) {
+            File propertiesFile = new File(driverPropertiesFile);
+            if (propertiesFile.exists()) {
+                LOG.fine(
+                        "Loading properties from the file:'" + driverPropertiesFile + "'"
+                );
+                FileInputStream inputStream = new FileInputStream(propertiesFile);
+                try {
+                    driverProperties.load(inputStream);
+                } finally {
+                    inputStream.close();
+                }
+            } else {
+                throw new RuntimeException("Can't open JDBC Driver specific properties from the file: '"
+                        + driverPropertiesFile + "'");
+            }
+        }
+
+
+        LOG.fine("Properties:");
+        for (Map.Entry<Object, Object> entry : driverProperties.entrySet()) {
+            if (entry.getKey().toString().toLowerCase().contains("password")) {
+                Scope.getCurrentScope().getLog(getClass()).fine("Key:'" + entry.getKey().toString() + "' Value:'**********'");
+            } else {
+                LOG.fine("Key:'" + entry.getKey().toString() + "' Value:'" + entry.getValue().toString() + "'");
+            }
+        }
+        return driverProperties;
     }
 
     private static class DatabaseComparator implements Comparator<Database> {

--- a/liquibase-core/src/main/java/liquibase/database/DatabaseFactory.java
+++ b/liquibase-core/src/main/java/liquibase/database/DatabaseFactory.java
@@ -11,6 +11,7 @@ import liquibase.util.StringUtil;
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
 import java.sql.Driver;
 import java.util.*;
 
@@ -156,6 +157,21 @@ public class DatabaseFactory {
                                              String driverPropertiesFile,
                                              String propertyProviderClass,
                                              ResourceAccessor resourceAccessor) throws DatabaseException {
+        Properties driverProperties;
+        try {
+            driverProperties = buildDriverProperties(username, password, driverPropertiesFile, propertyProviderClass);
+        } catch (Exception e) {
+            throw new DatabaseException(e);
+        }
+        return openConnection(url, username, driver, databaseClass, driverProperties, resourceAccessor);
+    }
+
+    public DatabaseConnection openConnection(String url,
+                                             String username,
+                                             String driver,
+                                             String databaseClass,
+                                             Properties driverProperties,
+                                             ResourceAccessor resourceAccessor) throws DatabaseException {
 
         if (url.startsWith("offline:")) {
             OfflineConnection offlineConnection = new OfflineConnection(url, resourceAccessor);
@@ -175,10 +191,8 @@ public class DatabaseFactory {
             Driver driverObject = loadDriver(selectedDriverClass);
 
             if (driverObject instanceof LiquibaseExtDriver) {
-                ((LiquibaseExtDriver)driverObject).setResourceAccessor(resourceAccessor);
+                ((LiquibaseExtDriver) driverObject).setResourceAccessor(resourceAccessor);
             }
-
-            Properties driverProperties = buildDriverProperties(username, password, driverPropertiesFile, propertyProviderClass);
 
             if(selectedDriverClass.contains("oracle")) {
               driverProperties.put("remarksReporting", "true");
@@ -250,36 +264,41 @@ public class DatabaseFactory {
         return driverObject;
     }
 
-    private Properties buildDriverProperties(String username, String password, String driverPropertiesFile, String propertyProviderClass) throws InstantiationException, IllegalAccessException, java.lang.reflect.InvocationTargetException, NoSuchMethodException, ClassNotFoundException, IOException {
+    private Properties buildDriverProperties(String username, String password, String driverPropertiesFile, String propertyProviderClass) {
         Properties driverProperties;
-        if (propertyProviderClass == null) {
-            driverProperties = new Properties();
-        } else {
-            driverProperties = (Properties) Class.forName(propertyProviderClass, true, Scope.getCurrentScope().getClassLoader()).getConstructor().newInstance();
-        }
-
-        if (username != null) {
-            driverProperties.put("user", username);
-        }
-        if (password != null) {
-            driverProperties.put("password", password);
-        }
-        if (null != driverPropertiesFile) {
-            File propertiesFile = new File(driverPropertiesFile);
-            if (propertiesFile.exists()) {
-                LOG.fine(
-                        "Loading properties from the file:'" + driverPropertiesFile + "'"
-                );
-                FileInputStream inputStream = new FileInputStream(propertiesFile);
-                try {
-                    driverProperties.load(inputStream);
-                } finally {
-                    inputStream.close();
-                }
+        try {
+            if (propertyProviderClass == null) {
+                driverProperties = new Properties();
             } else {
-                throw new RuntimeException("Can't open JDBC Driver specific properties from the file: '"
-                        + driverPropertiesFile + "'");
+                driverProperties = (Properties) Class.forName(propertyProviderClass, true, Scope.getCurrentScope().getClassLoader()).getConstructor().newInstance();
             }
+
+            if (username != null) {
+                driverProperties.put("user", username);
+            }
+            if (password != null) {
+                driverProperties.put("password", password);
+            }
+            if (null != driverPropertiesFile) {
+                File propertiesFile = new File(driverPropertiesFile);
+                if (propertiesFile.exists()) {
+                    LOG.fine(
+                            "Loading properties from the file:'" + driverPropertiesFile + "'"
+                    );
+                    FileInputStream inputStream = new FileInputStream(propertiesFile);
+                    try {
+                        driverProperties.load(inputStream);
+                    } finally {
+                        inputStream.close();
+                    }
+                } else {
+                    throw new RuntimeException("Can't open JDBC Driver specific properties from the file: '"
+                            + driverPropertiesFile + "'");
+                }
+            }
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException | ClassNotFoundException | IOException e) {
+            throw new RuntimeException("Exception opening JDBC Driver specific properties from the file: '"
+                    + driverPropertiesFile + "'", e);
         }
 
 

--- a/liquibase-core/src/main/java/liquibase/integration/ant/AbstractDatabaseDiffTask.java
+++ b/liquibase-core/src/main/java/liquibase/integration/ant/AbstractDatabaseDiffTask.java
@@ -21,7 +21,7 @@ public abstract class AbstractDatabaseDiffTask extends BaseLiquibaseTask {
     protected DiffResult getDiffResult() {
         Liquibase liquibase = getLiquibase();
         Database targetDatabase = liquibase.getDatabase();
-        Database referenceDatabase = createDatabaseFromType(referenceDatabaseType);
+        Database referenceDatabase = createDatabaseFromType(referenceDatabaseType, getResourceAccessor());
 
         CatalogAndSchema targetCatalogAndSchema = buildCatalogAndSchema(targetDatabase);
         CatalogAndSchema referenceCatalogAndSchema = buildCatalogAndSchema(referenceDatabase);

--- a/liquibase-core/src/main/java/liquibase/integration/ant/BaseLiquibaseTask.java
+++ b/liquibase-core/src/main/java/liquibase/integration/ant/BaseLiquibaseTask.java
@@ -63,7 +63,7 @@ public abstract class BaseLiquibaseTask extends Task {
         validateParameters();
         final Database[] database = {null};
         try {
-            ResourceAccessor resourceAccessor = createResourceAccessor(classLoader);
+            resourceAccessor = createResourceAccessor(classLoader);
             scopeValues.put(Scope.Attr.resourceAccessor.name(), resourceAccessor);
             scopeValues.put(Scope.Attr.classLoader.name(), classLoader);
 

--- a/liquibase-core/src/main/java/liquibase/integration/ant/BaseLiquibaseTask.java
+++ b/liquibase-core/src/main/java/liquibase/integration/ant/BaseLiquibaseTask.java
@@ -68,7 +68,7 @@ public abstract class BaseLiquibaseTask extends Task {
             scopeValues.put(Scope.Attr.classLoader.name(), classLoader);
 
             Scope.child(scopeValues, () -> {
-                database[0] = createDatabaseFromType(databaseType);
+                database[0] = createDatabaseFromType(databaseType, resourceAccessor);
                 liquibase = new Liquibase(getChangeLogFile(), resourceAccessor, database[0]);
                 if (changeLogParameters != null) {
                     changeLogParameters.applyParameters(liquibase);
@@ -95,11 +95,11 @@ public abstract class BaseLiquibaseTask extends Task {
     protected abstract void executeWithLiquibaseClassloader() throws BuildException;
 
     protected Database createDatabaseFromConfiguredDatabaseType() {
-        return createDatabaseFromType(databaseType);
+        return createDatabaseFromType(databaseType, getResourceAccessor());
     }
 
-    protected Database createDatabaseFromType(DatabaseType databaseType) {
-        return databaseType.createDatabase(classLoader);
+    protected Database createDatabaseFromType(DatabaseType databaseType, ResourceAccessor resourceAccessor) {
+        return databaseType.createDatabase(resourceAccessor);
     }
 
     protected Liquibase getLiquibase() {

--- a/liquibase-core/src/main/java/liquibase/integration/ant/BaseLiquibaseTask.java
+++ b/liquibase-core/src/main/java/liquibase/integration/ant/BaseLiquibaseTask.java
@@ -36,6 +36,7 @@ public abstract class BaseLiquibaseTask extends Task {
 
     private AntClassLoader classLoader;
     private Liquibase liquibase;
+    private ResourceAccessor resourceAccessor;
 
     private Path classpath;
     private DatabaseType databaseType;
@@ -105,6 +106,13 @@ public abstract class BaseLiquibaseTask extends Task {
         return liquibase;
     }
 
+    protected ResourceAccessor getResourceAccessor() {
+        if (resourceAccessor == null) {
+            throw new IllegalStateException("The ResourceAccessor has not been initialized. This usually means this " +
+                    "method has been called before the task's execute method has called.");
+        }
+        return resourceAccessor;
+    }
 
     /**
      * This method is designed to be overridden by subclasses when a change log is needed. By default it returns null.

--- a/liquibase-core/src/main/java/liquibase/integration/ant/type/DatabaseType.java
+++ b/liquibase-core/src/main/java/liquibase/integration/ant/type/DatabaseType.java
@@ -1,22 +1,14 @@
 package liquibase.integration.ant.type;
 
 import liquibase.database.Database;
-import liquibase.database.DatabaseConnection;
 import liquibase.database.DatabaseFactory;
-import liquibase.database.OfflineConnection;
-import liquibase.database.jvm.JdbcConnection;
 import liquibase.exception.DatabaseException;
-import liquibase.resource.ClassLoaderResourceAccessor;
+import liquibase.resource.ResourceAccessor;
 import org.apache.tools.ant.BuildException;
 import org.apache.tools.ant.Project;
 import org.apache.tools.ant.types.DataType;
 import org.apache.tools.ant.types.Reference;
-import org.apache.tools.ant.util.ClasspathUtils;
-import org.apache.tools.ant.util.LoaderUtils;
 
-import java.sql.Connection;
-import java.sql.Driver;
-import java.sql.SQLException;
 import java.util.Properties;
 
 public class DatabaseType extends DataType {
@@ -47,53 +39,33 @@ public class DatabaseType extends DataType {
         setProject(project);
     }
 
-    public Database createDatabase() {
-        ClassLoader contextClassLoader = LoaderUtils.getContextClassLoader();
-        return createDatabase(contextClassLoader);
-    }
-
-    public Database createDatabase(ClassLoader classLoader) {
+    public Database createDatabase(ResourceAccessor resourceAccessor) {
         logParameters();
         validateParameters();
         try {
             DatabaseFactory databaseFactory = DatabaseFactory.getInstance();
-            if(databaseClass != null) {
-                Database databaseInstance = (Database) ClasspathUtils.newInstance(databaseClass, classLoader, Database.class);
-                databaseFactory.register(databaseInstance);
+
+            Properties connectionProps = new Properties();
+            String connectionUserName = getUser();
+            if((connectionUserName != null) && !connectionUserName.isEmpty()) {
+                connectionProps.setProperty(USER_PROPERTY_NAME, connectionUserName);
+            }
+            String connectionPassword = getPassword();
+            if((connectionPassword != null) && !connectionPassword.isEmpty()) {
+                connectionProps.setProperty(PASSWORD_PROPERTY_NAME, connectionPassword);
+            }
+            ConnectionProperties dbConnectionProperties = getConnectionProperties();
+            if(dbConnectionProperties != null) {
+                connectionProps.putAll(dbConnectionProperties.buildProperties());
             }
 
-            DatabaseConnection jdbcConnection;
-            if (getUrl().startsWith("offline:")) {
-                jdbcConnection = new OfflineConnection(getUrl(), new ClassLoaderResourceAccessor(classLoader));
-            } else {
-                Driver jdbcDriver = (Driver) ClasspathUtils.newInstance(getDriver(), classLoader, Driver.class);
-                if(jdbcDriver == null) {
-                    throw new BuildException("Unable to create Liquibase Database instance. Could not instantiate the" +
-                     " JDBC driver.");
-                }
-                Properties connectionProps = new Properties();
-                String connectionUserName = getUser();
-                if((connectionUserName != null) && !connectionUserName.isEmpty()) {
-                    connectionProps.setProperty(USER_PROPERTY_NAME, connectionUserName);
-                }
-                String connectionPassword = getPassword();
-                if((connectionPassword != null) && !connectionPassword.isEmpty()) {
-                    connectionProps.setProperty(PASSWORD_PROPERTY_NAME, connectionPassword);
-                }
-                ConnectionProperties dbConnectionProperties = getConnectionProperties();
-                if(dbConnectionProperties != null) {
-                    connectionProps.putAll(dbConnectionProperties.buildProperties());
-                }
-
-                Connection connection = jdbcDriver.connect(getUrl(), connectionProps);
-                if(connection == null) {
-                    throw new BuildException("Unable to create Liquibase database instance. Could not connect to the " +
-                     "database.");
-                }
-                jdbcConnection = new JdbcConnection(connection);
-            }
-
-            Database database = databaseFactory.findCorrectDatabaseImplementation(jdbcConnection);
+            Database database = databaseFactory.openDatabase(
+                    getUrl(),
+                    getUser(),
+                    getDriver(),
+                    getDatabaseClass(),
+                    connectionProps,
+                    resourceAccessor);
 
             String schemaName = getDefaultSchemaName();
             if (schemaName != null) {
@@ -137,9 +109,6 @@ public class DatabaseType extends DataType {
             }
 
             return database;
-        } catch (SQLException e) {
-            throw new BuildException("Unable to create Liquibase database instance. A JDBC error occurred. " + e
-            .toString(), e);
         } catch (DatabaseException e) {
             throw new BuildException("Unable to create Liquibase database instance. " + e.toString(), e);
         }

--- a/liquibase-core/src/test/groovy/liquibase/change/core/DropAllForeignKeyConstraintsChangeTest.groovy
+++ b/liquibase-core/src/test/groovy/liquibase/change/core/DropAllForeignKeyConstraintsChangeTest.groovy
@@ -32,6 +32,8 @@ class DropAllForeignKeyConstraintsChangeTest extends Specification {
 
     @Unroll("SQL in #shortDbName for dropAllForeignKeys")
     def "GenerateStatements"() {
+        setup:
+        DatabaseFactory.reset()
         when:
         Database dbms = DatabaseFactory.getInstance().getDatabase(shortDbName)
         OfflineConnection conn = new OfflineConnection("offline:" + shortDbName, null)

--- a/liquibase-core/src/test/java/liquibase/database/DatabaseFactoryTest.java
+++ b/liquibase-core/src/test/java/liquibase/database/DatabaseFactoryTest.java
@@ -59,6 +59,21 @@ public class DatabaseFactoryTest {
     }
 
     @Test
+    public void openConnectionUsesDriverArgument() throws Exception {
+        DatabaseConnection dbConnection = databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", "org.h2.Driver", null, null, null, resourceAccessor);
+        assertThat(dbConnection, notNullValue());
+    }
+
+    @Test
+    public void openConnectionThrowsExceptionWhenDriverCannotBeFoundByUrl() throws Exception {
+        expectedException.expect(instanceOf(DatabaseException.class));
+        expectedException.expectCause(instanceOf(RuntimeException.class));
+        expectedException.expectMessage(containsString("Driver class was not specified and could not be determined from the url"));
+
+        databaseFactory.openConnection("not:a:driver", "", "", null, resourceAccessor);
+    }
+
+    @Test
     public void openConnectionLoadsDriverPropertiesFromGivenFile() throws Exception {
         File propsFile = temporaryFolder.newFile("db-factory-test-connection-props.properties");
         Properties expectedProps = new Properties();
@@ -88,13 +103,12 @@ public class DatabaseFactoryTest {
 
         databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", null, null, "unknown file", null, resourceAccessor);
     }
-
-    // Misc
-    // TODO: Delete this test
+    
     @Test
-    public void name() throws Exception {
+    public void openConnectionReturnsAConnection() throws Exception {
         DatabaseConnection databaseConnection = databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", null, resourceAccessor);
-        assertNotNull(databaseConnection);
+        assertThat(databaseConnection, notNullValue());
+        assertThat(databaseConnection.getDatabaseProductName(), equalTo("H2"));
     }
 
     /**

--- a/liquibase-core/src/test/java/liquibase/database/DatabaseFactoryTest.java
+++ b/liquibase-core/src/test/java/liquibase/database/DatabaseFactoryTest.java
@@ -35,6 +35,7 @@ public class DatabaseFactoryTest {
     @Before
     public void setUp() {
         resourceAccessor = mock(ResourceAccessor.class);
+        DatabaseFactory.reset();
         databaseFactory = DatabaseFactory.getInstance();
     }
 

--- a/liquibase-core/src/test/java/liquibase/database/DatabaseFactoryTest.java
+++ b/liquibase-core/src/test/java/liquibase/database/DatabaseFactoryTest.java
@@ -1,13 +1,110 @@
 package liquibase.database;
 
+import liquibase.exception.DatabaseException;
+import liquibase.resource.ResourceAccessor;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
 
+import java.io.File;
+import java.io.FileWriter;
+import java.util.Properties;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.notNullValue;
 import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.mock;
 
 public class DatabaseFactoryTest {
-    
+
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    private ResourceAccessor resourceAccessor;
+    private DatabaseFactory databaseFactory;
+
+    @Before
+    public void setUp() {
+        resourceAccessor = mock(ResourceAccessor.class);
+        databaseFactory = DatabaseFactory.getInstance();
+    }
+
+    @After
+    public void tearDown() {
+        temporaryFolder.delete();
+    }
+
     @Test
     public void getInstance() {
         assertNotNull(DatabaseFactory.getInstance());
+    }
+
+    // openConnection() method tests
+    @Test
+    public void openConnectionReturnsOfflineConnectionWhenUrlPrefixMatches() throws Exception {
+        String username = "sa";
+        DatabaseConnection dbConnection = databaseFactory.openConnection("offline:h2?param1=value1&aparam2=value2", username, "", null, resourceAccessor);
+        assertThat(dbConnection, notNullValue());
+        assertThat(dbConnection, instanceOf(OfflineConnection.class));
+        assertThat(dbConnection.getConnectionUserName(), equalTo(username));
+    }
+
+    @Test
+    public void openConnectionLoadsDriverPropertiesFromGivenFile() throws Exception {
+        File propsFile = temporaryFolder.newFile("db-factory-test-connection-props.properties");
+        Properties expectedProps = new Properties();
+        expectedProps.setProperty("param1", "value1");
+        expectedProps.setProperty("param2", "value2");
+        try (FileWriter writer = new FileWriter(propsFile)) {
+            expectedProps.store(writer, "connection properties");
+        }
+        String propsFilePath = propsFile.getAbsolutePath();
+
+        DatabaseConnection dbConnection = databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", null, null, propsFilePath, null, resourceAccessor);
+        assertThat(dbConnection, notNullValue());
+        // TODO: Figure out how to assert the properties are loaded
+    }
+
+    @Test
+    public void openConnectionCreatesCustomPropertyProviderClassWhenGiven() throws Exception {
+        DatabaseConnection databaseConnection = databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", "liquibase.database.DatabaseFactoryTest$CustomProperties", resourceAccessor);
+        assertThat(databaseConnection, notNullValue());
+    }
+
+    @Test
+    public void openConnectionThrowsRuntimeExceptionWhenDriverPropertiesFileNotFound() throws Exception {
+        expectedException.expect(instanceOf(DatabaseException.class));
+        expectedException.expectCause(instanceOf(RuntimeException.class));
+        expectedException.expectMessage(containsString("Can't open JDBC Driver specific properties from the file"));
+
+        databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", null, null, "unknown file", null, resourceAccessor);
+    }
+
+    // Misc
+    // TODO: Delete this test
+    @Test
+    public void name() throws Exception {
+        DatabaseConnection databaseConnection = databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", null, resourceAccessor);
+        assertNotNull(databaseConnection);
+    }
+
+    /**
+     * Simple type to test the propertyProviderClass
+     */
+    @SuppressWarnings({"unused", "RedundantSuppression"})
+    private static class CustomProperties extends Properties {
+        // Default constructor needed for reflective construction
+        public CustomProperties() {
+            super();
+        }
     }
 }

--- a/liquibase-core/src/test/java/liquibase/database/DatabaseFactoryTest.java
+++ b/liquibase-core/src/test/java/liquibase/database/DatabaseFactoryTest.java
@@ -62,6 +62,7 @@ public class DatabaseFactoryTest {
     public void openConnectionUsesDriverArgument() throws Exception {
         DatabaseConnection dbConnection = databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", "org.h2.Driver", null, null, null, resourceAccessor);
         assertThat(dbConnection, notNullValue());
+        assertThat(dbConnection.getDatabaseProductName(), equalTo("H2"));
     }
 
     @Test
@@ -71,6 +72,13 @@ public class DatabaseFactoryTest {
         expectedException.expectMessage(containsString("Driver class was not specified and could not be determined from the url"));
 
         databaseFactory.openConnection("not:a:driver", "", "", null, resourceAccessor);
+    }
+
+    @Test
+    public void openConnectionLoadsGivenDatabaseClass() throws Exception {
+        DatabaseConnection dbConnection = databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", null, "liquibase.database.core.H2Database", null, null, resourceAccessor);
+        assertThat(dbConnection, notNullValue());
+        assertThat(dbConnection.getDatabaseProductName(), equalTo("H2"));
     }
 
     @Test
@@ -86,13 +94,15 @@ public class DatabaseFactoryTest {
 
         DatabaseConnection dbConnection = databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", null, null, propsFilePath, null, resourceAccessor);
         assertThat(dbConnection, notNullValue());
+        assertThat(dbConnection.getDatabaseProductName(), equalTo("H2"));
         // TODO: Figure out how to assert the properties are loaded
     }
 
     @Test
     public void openConnectionCreatesCustomPropertyProviderClassWhenGiven() throws Exception {
-        DatabaseConnection databaseConnection = databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", "liquibase.database.DatabaseFactoryTest$CustomProperties", resourceAccessor);
-        assertThat(databaseConnection, notNullValue());
+        DatabaseConnection dbConnection = databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", "liquibase.database.DatabaseFactoryTest$CustomProperties", resourceAccessor);
+        assertThat(dbConnection, notNullValue());
+        assertThat(dbConnection.getDatabaseProductName(), equalTo("H2"));
     }
 
     @Test
@@ -106,9 +116,9 @@ public class DatabaseFactoryTest {
     
     @Test
     public void openConnectionReturnsAConnection() throws Exception {
-        DatabaseConnection databaseConnection = databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", null, resourceAccessor);
-        assertThat(databaseConnection, notNullValue());
-        assertThat(databaseConnection.getDatabaseProductName(), equalTo("H2"));
+        DatabaseConnection dbConnection = databaseFactory.openConnection("jdbc:h2:mem:DatabaseFactoryTest", "sa", "", null, resourceAccessor);
+        assertThat(dbConnection, notNullValue());
+        assertThat(dbConnection.getDatabaseProductName(), equalTo("H2"));
     }
 
     /**

